src/script/interpreter.cpp

```diff
@@ -907,10 +917,61 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
                     CScript scriptCode(pbegincodehash, pend);
 
                     // Drop the signatures, since there's no way for a signature to sign itself
+                    //
+                    // When SCRIPT_VERIFY_REQUIRE_VALID_SIGSS is in effect, a requirement
+                    // is added that when CHECKMULTISIG fails, ALL signatures must have
+                    // been empty. So that we can later verify this requirement, we record
+                    // whether signatures are provided, valid or otherwise.
+                    bool empty_sigs;
                     for (int k = 0; k < nSigsCount; k++)
                     {
                         valtype& vchSig = stacktop(-isig-k);
                         scriptCode.FindAndDelete(CScript(vchSig));
+                        // If the first signature is empty, then they all must be empty.
+                        // If the first signature is non-empty, then they all must be non-empty.
+                        if (k == 0) {
+                            empty_sigs = vchSig.empty();
+                        } else {
+                            if ( ( empty_sigs && !vchSig.empty()) ||
+                                 (!empty_sigs &&  vchSig.empty()) )
```

Comment:
It initially appears like you are using `empty_sigs` uninitialized, until you read more carefully. Isn't it more straightforward to just do
```C++
size_t empty_sigs = 0;
```
then
```C++
empty_sigs += vchSig.empty();
```
and after loop
```C++
if (empty_sigs > 0 && empty_sigs < nSigsCount) {
    return set_error(serror, SCRIPT_ERR_FAILED_SIGNATURE_CHECK);
}
```

Reply:
Yeah that's much clearer. I adopted that with a few belts-and-suspenders modifications.

Reply:
Yeah that's much clearer. I adopted that with a few belts-and-suspenders modifications.